### ISSUE TITLE
Running Rank 1 on TREC-DL Datasets

### ISSUE THREAD

**Issue by ChuanMeng** [orionw/rank1#5]
Hi Orion,

Thank you for your excellent work on Rank 1!

Could you please give me some instructions on running Rank1 on the TREC-DL series datasets, such as TREC-DL 19 and 20? Also in the prompt.py, it seems that it doesn't contain the prompt used for TREC-DL?

Thank you in advance!

Best regards,
Chuan

**Comment by orionw** [comment]
Hi @ChuanMeng, great question! 

Surprisingly, we still don't have the DL series on MTEB (I just opened an issue: https://github.com/embeddings-benchmark/mteb/issues/3348). 

For the Rank1 paper, I evaluated it using the [`rank_llm`](https://github.com/castorini/rank_llm) package, but I didn't have enough time to make a PR to their repo. Here's the code file for the model implementation: https://gist.github.com/orionw/85ed93c8f4d806c881a8ee25c89514ee

It has the prompt used in that gist (very basic). If you wanted to actually use this in their package with this though, you have to add it as a new model in the `pointwise` folder and then make sure it's imported in the correct places, so it is a bit of a pain.  Sorry about that! Honestly it might be easier to just add the TREC dataset to MTEB and evaluate there since all that requires is cloning [this dataset](https://huggingface.co/datasets/mteb/msmarco) and replacing the qrels and queries files. 

Sorry I didn't make that easier and let me know if you have any more questions!

**Comment by ChuanMeng** [comment]
Thank you Orion for your quick response and for sharing the code; it’s very helpful.

I have three follow-up questions:
1. The code provides both vllm_batched and transformers implementations. Do you only use vllm_batched?

2. Do you include titles for passages in MS MARCO?

3. When I run the code you shared within rank_llm, I noticed that the text output by the model (all_final_texts in the code) does not include `</think>` or `true/false`, but rather reasoning-style text such as “.`.. so this passage isn’t relevant to the query.`” According to the paper, should the text output be "`.. so this passage isn’t relevant to the query. </think> false`"? 

I look forward to your reply.

Best regards,
Chuan

**Comment by orionw** [comment]
> The code provides both vllm_batched and transformers implementations. Do you only use vllm_batched?

Yes, although vLLM doesn't return all the logits, so in some cases it has to fall back to transformers when the </think> tokens don't appear in the top 20.  That's why the code is so complex sadly.

> Do you include titles for passages in MS MARCO?

I am not 100% sure on this, that would depend on what `rank_llm` does by default. I was able to reproduce their scores with their models, so whatever that would entail.

> When I run the code you shared within rank_llm, I noticed that the text output by the model (all_final_texts in the code) does not include </think> or true/false, but rather reasoning-style text such as “... so this passage isn’t relevant to the query.” According to the paper, should the text output be ".. so this passage isn’t relevant to the query. </think> false"?

This is an artifact of vLLM if I understand you correctly, it doesn't output the stop tokens in the text (so it will strip out the `</think> true` which is one of the stop tokens) but you can get it back with the logits IIRC.

**Comment by ChuanMeng** [comment]
I have two follow-up questions based on your answers.

First, could you please clarify what you mean by “it has to fall back to transformers when the tokens don't appear in the top 20”?
According to your code, if the "true" or "false" tokens are not in the top 20 of the last token in the generated text (see the code below), the function `_fix_incomplete_responses ` will be executed. However, `_fix_incomplete_responses` only uses vLLM and does not rely on transformers at all.

> if self.true_token not in final_logits or self.false_token not in final_logits:
>     incomplete_prompts.append(prompts[i])
>     incomplete_texts.append(text)
>     incomplete_indices.append(i)
>     continue


Second, it seems that whether we invoke `_fix_incomplete_responses` depends on whether the "true" or "false" tokens are not among the top 20 logprobs of the last token in the generated text. However, sometimes the model may generate text that does not follow the expected format — for example, missing an ending like "</think> true" or "</think> false". In such cases, even if "true" or "false" are within the top 20 logprobs, the actual last generated token may still not be "true" or "false".
Would it make sense to modify the condition so that `_fix_incomplete_responses` is triggered when the generated text does not end with either "</think> true" or "</think> false"? (see the code below)

>  if not (text.endswith("</think> true") or text.endswith("</think> false")):
>         incomplete_prompts.append(prompts[i])
>         incomplete_texts.append(text)
>         incomplete_indices.append(i)
>         continue

I look forward to your reply.

Best regards,
Chuan

**Comment by orionw** [comment]
> First, could you please clarify what you mean by “it has to fall back to transformers when the tokens don't appear in the top 20”?

Ah you are right, sorry! I am misremembering my own code. I initially had the fallback to transformers but forgot I figured out a way to have it work in vLLM the whole time.

> Would it make sense to modify the condition so that _fix_incomplete_responses is triggered when the generated text does not end with either " true" or " false"?

Maybe I'm misunderstanding your question but it needs to end with `</think> true` or `</think> false` given the training data format. You're welcome to change how it does the fall back how you want it though, I imagine there are many things one could do and this is only one approach.

**Comment by ChuanMeng** [comment]
Thank you Orion for your quick response. Much appreciated. It solved my questions :-)
